ethcore/src/trace/executive_tracer.rs

```diff
@@ -17,159 +17,166 @@
-		trace.trace_address.push_front(current_subtrace_index);
-	}
-	traces
-}
-
-#[test]
-fn should_prefix_address_properly() {
-	use super::trace::{Action, Res, Suicide};
-
-	let f = |v: Vec<usize>| FlatTrace {
-		action: Action::Suicide(Suicide {
-			address: Default::default(),
-			balance: Default::default(),
-			refund_address: Default::default(),
-		}),
-		result: Res::None,
-		subtraces: 0,
-		trace_address: v.into_iter().collect(),
-	};
-	let t = vec![vec![], vec![0], vec![0, 0], vec![0], vec![], vec![], vec![0], vec![]].into_iter().map(&f).collect();
-	let t = prefix_subtrace_addresses(t);
-	assert_eq!(t, vec![vec![0], vec![0, 0], vec![0, 0, 0], vec![0, 0], vec![1], vec![2], vec![2, 0], vec![3]].into_iter().map(&f).collect::<Vec<_>>());
+	index_stack: Vec<usize>,
+	vecindex_stack: Vec<usize>,
+	sublen_stack: Vec<usize>,
+	skip_one: bool,
 }
 
 impl Tracer for ExecutiveTracer {
 	type Output = FlatTrace;
 
-	fn prepare_trace_call(&self, params: &ActionParams) -> Option<Call> {
-		Some(Call::from(params.clone()))
-	}
+	fn prepare_trace_call(&mut self, params: &ActionParams, depth: usize, is_builtin: bool) {
+		if depth != 0 && is_builtin && params.value.value() == U256::zero() {
+			self.skip_one = true;
+			return;
+		}
 
-	fn prepare_trace_create(&self, params: &ActionParams) -> Option<Create> {
-		Some(Create::from(params.clone()))
-	}
+		if let Some(parentlen) = self.sublen_stack.last_mut() {
+			*parentlen += 1;
+		}
 
-	fn trace_call(&mut self, call: Option<Call>, gas_used: U256, output: &[u8], subs: Vec<FlatTrace>) {
 		let trace = FlatTrace {
-			trace_address: Default::default(),
-			subtraces: top_level_subtraces(&subs),
-			action: Action::Call(call.expect("self.prepare_trace_call().is_some(): so we must be tracing: qed")),
+			trace_address: self.index_stack.clone(),
+			subtraces: self.sublen_stack.last().cloned().unwrap_or(0),
+			action: Action::Call(Call::from(params.clone())),
 			result: Res::Call(CallResult {
-				gas_used: gas_used,
-				output: output.into()
+				gas_used: U256::zero(),
+				output: Vec::new()
 			}),
 		};
-		debug!(target: "trace", "Traced call {:?}", trace);
+		self.vecindex_stack.push(self.traces.len());
 		self.traces.push(trace);
-		self.traces.extend(prefix_subtrace_addresses(subs));
+		self.index_stack.push(0);
+		self.sublen_stack.push(0);
 	}
 
-	fn trace_create(&mut self, create: Option<Create>, gas_used: U256, code: &[u8], address: Address, subs: Vec<FlatTrace>) {
+	fn prepare_trace_create(&mut self, params: &ActionParams) {
+		if let Some(parentlen) = self.sublen_stack.last_mut() {
+			*parentlen += 1;
+		}
+
 		let trace = FlatTrace {
-			subtraces: top_level_subtraces(&subs),
-			action: Action::Create(create.expect("self.prepare_trace_create().is_some(): so we must be tracing: qed")),
+			trace_address: self.index_stack.clone(),
+			subtraces: self.sublen_stack.last().cloned().unwrap_or(0),
+			action: Action::Create(Create::from(params.clone())),
 			result: Res::Create(CreateResult {
```

Comment:
maybe it's worth introducing a third variant? or redesigning the implementation to avoid dummy data creation?

Reply:
I think this may be okay. `U256::zero()`, `Vec::new()`, `Address::default()` all would just zero out the fields, and wouldn't allocate anything on the heap.

* From performance's perspective, it wouldn't make any difference if we do dummy data creation or use `Option`, they would both just zeroing out the fields.
* From clarity's perspective, this dummy data creation is restraint in `ExecutiveTracer` and doesn't add any complexity outside of it.